File: Lab 3/merge.py
def mergeSort(arr, left, right):
    if left < right:
        mid = (left + right) // 2
        
        # Print current subarray being divided
        print("Dividing:", end=" ")
        for i in range(left, right + 1):
            print(arr[i], end=" ")
        print()

        # Recursively divide the array into two halves
        mergeSort(arr, left, mid)
        mergeSort(arr, mid + 1, right)

        # Merge the two halves
        n1 = mid - left + 1
        n2 = right - mid
        leftArr = arr[left:left + n1]
        rightArr = arr[mid + 1:mid + 1 + n2]

        # Print left and right arrays before merging
        print("Left Array:", end=" ")
        for i in range(n1):
            print(leftArr[i], end=" ")
        print()

        print("Right Array:", end=" ")
        for i in range(n2):
            print(rightArr[i], end=" ")
        print()

        # Merge the two temporary arrays back into arr[]
        i = 0
        j = 0
        k = left
        while i < n1 and j < n2:
            if leftArr[i] <= rightArr[j]:
                arr[k] = leftArr[i]
                i += 1
            else:
                arr[k] = rightArr[j]
                j += 1
            k += 1

        # Copy remaining elements of leftArr[], if any
        while i < n1:
            arr[k] = leftArr[i]
            i += 1
            k += 1

        # Copy remaining elements of rightArr[], if any
        while j < n2:
            arr[k] = rightArr[j]
            j += 1
            k += 1

        # Print the merged array
        print("Merged Array:", end=" ")
        for i in range(left, right + 1):
            print(arr[i], end=" ")
        print()


def main():
    # Input the size of the array
    n = int(input("Enter the number of elements in the array: "))

    # Input the elements of the array
    arr = []
    print("Enter the elements of the array: ")
    for i in range(n):
        arr.append(int(input()))

    # Perform Merge Sort
    mergeSort(arr, 0, n - 1)

    # Output the sorted array
    print("Sorted array:", end=" ")
    for num in arr:
        print(num, end=" ")
    print()


if __name__ == "__main__":
    main()

File: Lab 3/merge.c
#include <stdio.h>

void conquer(int arr[], int left, int mid, int right) {
    int n1 = mid - left + 1, n2 = right - mid;
    int leftArr[n1], rightArr[n2];
    
    for (int i = 0; i < n1; i++) leftArr[i] = arr[left + i];
    for (int j = 0; j < n2; j++) rightArr[j] = arr[mid + 1 + j];
    
    printf("Left Array: ");
    for (int i = 0; i < n1; i++) printf("%d ", leftArr[i]);
    printf("\n");
    
    printf("Right Array: ");
    for (int j = 0; j < n2; j++) printf("%d ", rightArr[j]);
    printf("\n");
    
    int i = 0, j = 0, k = left;
    while (i < n1 && j < n2) {
        arr[k++] = (leftArr[i] <= rightArr[j]) ? leftArr[i++] : rightArr[j++];
    }
    while (i < n1) arr[k++] = leftArr[i++];
    while (j < n2) arr[k++] = rightArr[j++];
    
    printf("Merged Array: ");
    for (int i = left; i <= right; i++) printf("%d ", arr[i]);
    printf("\n");
}

void divide(int arr[], int left, int right) {
    if (left < right) {
        int mid = (left + right) / 2;
        
        printf("Dividing: ");
        for (int i = left; i <= right; i++) printf("%d ", arr[i]);
        printf("\n");
        
        divide(arr, left, mid);
        divide(arr, mid + 1, right);
        conquer(arr, left, mid, right);
    }
}

int main() {
    int n;
    printf("Enter the number of elements in the array: ");
    scanf("%d", &n);
    
    int arr[n];
    printf("Enter the elements of the array: ");
    for (int i = 0; i < n; i++) {
        scanf("%d", &arr[i]);
    }
    
    divide(arr, 0, n - 1);
    
    printf("Sorted array: ");
    for (int i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");
    
    return 0;
}

